### DISCUSSION TITLE
Update action hook is called BEFORE updating when used in combination with recursive relationships and MySQL

### DISCUSSION THREAD

**Discussion by christianrr** [directus/directus#10484] [Archive]
We have the following setup:

- Recursive O2M-relationship with tree view interface configured
- Now we're using tree view interface to change some child
- If we then query the item in update action hook (= AFTER hook), ItemsService returns the values BEFORE changing (it's the same if we query the database directly with knex)

Bug does happen with MySQL adapter, does not happen with SqLite.
Bug is also happening with O2M interface, so it might be related with the recursive relationship - maybe update order?

Everything is working as expected for "normal" O2M relationships...

### To Reproduce

- Create Table `pages`
- Create Field `title`
- Create Field `subpages`-> Tree view (relationship to `pages`, fk `page_id`, display template `title`)
- Add hook as below
- Add items (min. 1 level deep), e.g. `Root` -> `Child 1`
- Save items
- Go to `Root` element und use tree view interface to change `Child 1`'s title to `Child1 - changed`
- Save items
- Hook still returns `Child1` instead of changed title (see below)

```
    action('pages.items.update', async ({ keys }, { schema, accountability }) => {
        const itemsService = new ItemsService('pages', {
            schema: schema,
            accountability: accountability
        });

        for(let i=0; i<keys.length; i++) {
            let key = keys[i];
            
            const item = await itemsService.readOne(key, {
                fields: ["id", "title"],
            });
            
            console.log('changed item:', item);
        }
    });
```

Hook update (title should be "Child1 - changed")

```
changed item: { id: 2, title: 'Child 1' }
changed item: { id: 1, title: 'Root' }
```

**Comment by rijkvanzanten** [comment]
Heya! Converting this to a discussion, as I believe it's mostly a discussion on how it works/should work rather than a specific bug in Directus.

As of right now, the hooks are fired based on when the actual data point(s) is(/are) being created. When using a one-to-many it goes "top down", for many-to-one it's "bottom up". This is done because you need the primary keys on the parent/child to inject the full row. For example, lets say you have "authors" and "articles", where each author (one) can have multiple articles (many). When inserting:

```
name: Rijk
articles:
  - Hello!
  - Another article
```

The author is created first (which triggers the "items.create" hook for authors), after which the "articles" are created with the author_id correctly set.

From a many to one perspective, a similar thing happens:

```
title: Hello!
author:
  name: Rijk
```

In this case, the nested author is saved first, as the parent articles record needs to have the nested primary key to store in the m2o field.

When the data is a recursive relationship, the same concept still applies. This does mean that the hook can/will fire multiple times.

Secondly, nested relational updates are fired in a single database _transaction_. This does mean that queries fired _during_ the transaction that aren't _using_ the transaction might not see the data (yet), as the transaction query hasn't been committed yet. You can actually have the ItemsService in your hook example use the current transaction, to make sure you have the most up to date state of the current execution. The context parameter of the hook actually gets a copy of the database object used internally, which is either the root database connection, or the current transaction query.

```js
    action('pages.items.update', async ({ keys }, { schema, accountability, database }) => {
        const itemsService = new ItemsService('pages', {
            schema: schema,
            accountability: accountability,
            knex: database, // <<<<<<
        });

        for(let i=0; i<keys.length; i++) {
            let key = keys[i];
            
            const item = await itemsService.readOne(key, {
                fields: ["id", "title"],
            });
            
            console.log('changed item:', item);
        }
    });
```

Lemme know if all of that makes sense 🙂 Happy to keep discussing how it works and more importantly how we can improve things.

---

Also, all this information would be super useful to have in the docs (cc @benhaynes) though I'm not entirely sure _where_. It's more advanced hooks usage, but therefore not less important information 🤔 I wouldn't want to over-complicate the hooks documentation with these more advanced in-depth explanations, but I also don't want to omit this information. Any thoughts on a good spot to add these more in-depth details @benhaynes?

**Reply by christianrr** [reply]
Ok, thanks - get the internals better now. I think my problem here is mainly the single database transaction. 

I've tried to add the database instance from context into the ItemsService in the hook as instructed:

```
action('pages.items.update', async ({ keys }, { schema, accountability }) => {
        const itemsService = new ItemsService('pages', {
            schema: schema,
            accountability: accountability,
            knex: database
        });
       ...
    });
```

From your explanation, I would expect that if I pass in the database and the update hook for the nested item is fired first, I would get the latest data from the current transaction. However, this is not the case (see output below).

And that's also the main issue IMO, because at the moment when the nested item (id: 2) is changing, there is no possibility to get the latest data of that particular object. Neither there is any information that this change is part of a larger transaction, which will involve the parent (id: 1). If I would know, I could postpone the handling for the nested item until the parent is changed.

```
changed item: { id: 2, title: 'Nested Item', subpages: [] }
changed item: {
  id: 1,
  title: 'Root',
  subpages: [ { id: 2, title: '**Nested Item-Changed**' } ]
}
```

I hope it's clear what I mean - we are digging quite deep here 😄 Thanks for helping out.

**Reply by rijkvanzanten** [reply]
Yeah that makes total sense! 

We're currently already tracking / storing the "parent revision" under directus_revisions, so I'm wondering if we could extract that logic, and pass that on to the hook under the meta information. That way, you'd be able to see a "this event is fired as a child event for item X" 🤔

**Reply by christianrr** [reply]
Yeah, that would be indeed helpful. 

However, if I would wait for the root change to happen in my use case, it would be necessary to know which (nested) item changes were involved (somehow the way backwards).

In general, it would be often helpful to know the deltas in action hooks (similar to input in filter hook). Of course it is possible to extract this information via directus_revisions, but this is kind of complicated in action hooks...

**Reply by rijkvanzanten** [reply]
Another angle could be to have a hook that fires _before_ the relational insert / transaction starts, which has the full nested object as the payload 🤔
